sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/serialization/CoderBasedKafkaSerializer.java

```diff
@@ -20,29 +20,79 @@
 
 import static com.google.common.base.Preconditions.checkNotNull;
 
+import com.google.common.collect.ImmutableMap;
+
+import java.io.InputStream;
 import java.util.Map;
 import javax.annotation.Nullable;
+
+import org.apache.beam.sdk.coders.AvroCoder;
 import org.apache.beam.sdk.coders.Coder;
 import org.apache.beam.sdk.coders.CoderException;
+import org.apache.beam.sdk.io.kafka.KafkaIO;
 import org.apache.beam.sdk.util.CoderUtils;
+import org.apache.kafka.common.serialization.Deserializer;
 import org.apache.kafka.common.serialization.Serializer;
 
 /**
- * Implements Kafka's {@link Serializer} with a {@link Coder}.
+ * Implements a Kafka {@link Deserializer} and {@link Serializer} with a {@link Coder}.
```

Comment:
I suggest to point-blank delete this class. It's really useful only for handling Avro and Proto messages; examples in the tests which use it for integers are particularly misleading.

Let's figure out a good way to simplify dealing with Avro and Proto keys/values separately (not in this PR).